wander-embedded-sdk/src/WanderEmbedded.ts

```diff
@@ -0,0 +1,97 @@
+import { setupWalletSDK } from "wallet-api/wallet-sdk.es.js";
+import { WanderEmbeddedOptions } from "./types";
+import { WanderButton } from "./components/WanderButton";
+import { WanderIframe } from "./components/WanderIframe";
+import {
+  IncomingMessage,
+  IncomingMessageId,
+  IncomingResizeMessageData
+} from "./types/messages";
+import { UIComponents } from "./types/embedded";
+
+export class WanderEmbedded {
+  private components: UIComponents;
+  private readonly DEFAULT_IFRAME_SRC = "http://localhost:5174/";
+  private options: WanderEmbeddedOptions;
+
+  constructor(options?: WanderEmbeddedOptions) {
+    this.options = options || {};
+    this.components = this.initializeComponents(options);
+  }
+
+  private initializeComponents(options?: WanderEmbeddedOptions): UIComponents {
+    const iframe = new WanderIframe({
+      src: this.DEFAULT_IFRAME_SRC,
+      onMessage: (message) => this.handleIframeMessage(message),
+      iframeRef: options?.iframeRef,
+      iframeStyles: options?.iframeStyles
+    });
+
+    if (!options?.iframeRef) {
+      document.body.appendChild(iframe.getElement());
+    }
+    this.initializeWalletShim(iframe.getElement());
+
+    if (options?.buttonStyles !== "none") {
+      const container = this.createContainer();
+
+      const button = new WanderButton({
+        buttonStyles: options?.buttonStyles,
+        onClick: () => this.open(),
+        logo: options?.logo,
+        balance: options?.balance
+      });
+      container.appendChild(button.getElement());
+      document.body.appendChild(container);
+
+      return { container, button, iframe };
+    }
+
+    return { iframe };
+  }
+
+  private createContainer(): HTMLDivElement {
+    const container = document.createElement("div");
+    container.id = "wander-embedded-container";
+    return container;
+  }
+
+  private handleIframeMessage(
+    message: IncomingMessage<IncomingMessageId>
+  ): void {
+    switch (message.type) {
+      case "embedded_close":
+        this.components.iframe.hide();
+        this.options.onClose?.();
+        break;
+      case "embedded_resize":
+        this.components.iframe.resize(
+          message.data as IncomingResizeMessageData
+        );
+        this.components.iframe.show();
+
+        this.options.onOpen?.();
```

Comment:
We need to add a `isOpen` property and only call `onOpen` when `isOpen` goes from `false` to `true`.

Reply:
you mean something like below?

```ts
case "embedded_resize":
        this.components.iframe.resize(
          message.data as IncomingResizeMessageData
        );
        this.components.iframe.show();

        if (!this.isOpen) {
          this.isOpen = true;
          this.options.onOpen?.();
        }
        this.options.onResize?.(message.data as IncomingResizeMessageData);
        break;
```

Reply:
I would also avoid calling `this.components.iframe.show()` if it's already being shown:

```
case "embedded_resize":
        this.components.iframe.resize(
          message.data as IncomingResizeMessageData
        );

        if (!this.isOpen) {
          this.isOpen = true;
          this.components.iframe.show();
          this.options.onOpen?.();
        }

        this.options.onResize?.(message.data as IncomingResizeMessageData);
        break;
```